common/logger/src/struct_log.rs

```diff
@@ -0,0 +1,211 @@
+use chrono::Utc;
+use log;
+use serde_json;
+use std::io::Write as IoWrite;
+
+use serde::Serialize;
+use serde_json::Value;
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        mpsc::{self, Receiver, SyncSender},
+    },
+    thread,
+};
+
+pub trait StructLogSink: Sync {
+    fn send(&self, entry: StructuredLogEntry);
+}
+
+// This is poor's man AtomicReference from crossbeam
+// It have few unsafe lines, but does not require extra dependency
+static NOP: NopStructLog = NopStructLog {};
+static mut STRUCT_LOGGER: &'static dyn StructLogSink = &NOP;
+static STRUCT_LOGGER_STATE: AtomicUsize = AtomicUsize::new(UNINITIALIZED);
+const UNINITIALIZED: usize = 0;
+
+const INITIALIZING: usize = 1;
+const INITIALIZED: usize = 2;
+
+#[derive(Default, Serialize)]
+pub struct StructuredLogEntry {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    text: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pattern: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    module: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    location: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    git_rev: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    timestamp: Option<String>,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    data: HashMap<&'static str, Value>,
+}
+
+#[must_use = "use StructuredLogEntry::send to send structured log"]
+impl StructuredLogEntry {
+    pub fn new_unnamed() -> Self {
+        let mut ret = Self::default();
+        ret.timestamp = Some(Utc::now().format("%F %T").to_string());
+        ret
+    }
+
+    pub fn new_named(name: &'static str) -> Self {
+        let mut ret = Self::default();
+        ret.name = Some(name);
+        ret.timestamp = Some(Utc::now().format("%F %T").to_string());
+        ret
+    }
+
+    pub fn text(&mut self, text: String) -> &mut Self {
```

Comment:
I've seen several different ways of implementing setters in Rust, how did you land on this one?
Versus
fn set_*(mut self, ...) -> Self
fn set_*(&mut self, ...) -> &self
or some variant thereof...

Reply:
This one seem to be most flexible.
`set(mut self)` would require to use chaining or re-assign variables constantly